_posts/2021-12-13-kafka-segment-retention.md

```diff
@@ -0,0 +1,208 @@
+
+Anyway, even when we think that finally the retention time is evaluated after the last record, it could be still there!
+The reason is related to an Apache Kafka thread which runs periodically on the broker to clean records and checking which closed segment can be deleted. The `log.retention.check.interval.ms` is the parameter to configure the period for this thread to run and it defaults to 5 minutes.
+Depending on when the last record is appended in the segment, so the segment is closed and when the thread run the last time, it could miss the 10 minutes deadline of the retention and so deleting the closed segment later, so maybe up to almost 5 minutes delay.
+
+In our example, it could happen that the first record in the segment could live up to almost 24 minutes!
+
+Assuming that at some point the clean thread runs and verify that a closed segment can be deleted, initially it just change extensions of corresponding files by adding the `.deleted` extension but not actually deleting the segment from the file system.
+There is another broker parameter `log.segment.delete.delay.ms` (1 minute by default) which defines when the file will be actually cancelled from the file system when it's marked as "deleted".
+
+![Total Retention Time](/assets/images/posts/2021-12-13-total-retention-time.png)
+
+Coming back to the canary example again and adding the delay on deletion, the first record in a segment is still alive after 25 minutes!
+It is really longer than the 10 minutes expectation, isn't it?
+
+From the above, it is clear that the retention mechanism doesn't really match the initial expectation.
+In reality, a record could live even longer than our 10 minutes depending on a lot of other configuration and internal mechanics of the broker.
+The usual retention limits set by using `log.retention.ms` defines a kind of lower bound; Kafka guarantees it will not delete any records that have an age less than the one specified, but any older records could get deleted at any time in the future depending on different settings.
+
+It is also worth to mention how it has an impact on the consumer side.
+Consumers don't get records from closed segment as well as from deleted segment even if they are just marked as "deleted" but not actually removed from the file system.
+This is true even when a consumer starts to read the partition from the beginning.
+Longer retention won't have a direct impact on the consumers but more on the disk usage.
+
+It is worth mentioning that the overall retention mechanism works this way when the `log.cleanup.policy` is set to `delete`.
+For `compact`ed topics, it is completely different as you can read about on the official Strimzi documentation [here](https://strimzi.io/docs/operators/latest/using.html#removing_log_data_with_cleanup_policies).
```

Comment:
Maybe this is something you can mention early -> that this applies to topics with `delete` policy and not `compact`. When you do it only at the end, it might confuse the users while reading the post.

Reply:
I am not sure about it ... I mean. How could I start the entire paragraph by saying "everything I am going to explain is valid for delete policy and not for the compact one"? Isn't it better just leaving it out until the end as it is now?
Or at least what could be the better way to say it earlier?

Reply:
Well, your blog post, your call I guess. But while reading it I was wondering all the time about how does it work with compacting and if you are going to explain it. So it felt a bit weird and disappointing at the end. I thought knowing it upfront would make me think about it a bit less. But as I said, it is your call at the end, so do it as you feel.

Reply:
@PaulRMellor any suggestion where to put this sentence? Tbh I think it's the right place but having a suggestion could help me to understand if it's better somewhere else.

Reply:
I've suggested a compaction note at the start. If you're happy with that, you can remove this.